### ISSUE TITLE
How to change background color of BottomNavigationBarItem(child) when user pressed? Not change all of navbar color

### ISSUE THREAD

**Issue by rrifafauzikomara** [flutter/flutter#34788]
I want every menu that I click will change color for each menu that is clicked, not for the entire color background of the nav bar. 

Like the example below:
[Here](https://i.stack.imgur.com/dXGaz.png)

When a user clicks on the "Near Me" menu then the box on the menu will change color. But for the background/color box on the other menu, it is still black. Similarly, clicking on the other menus will change color to red and will not change the color of the menu that is not clicked (still black)

This my Code:
```
int _page = 0;
final List<Widget> _children = [
    MainPage(),
    MainPage(),
    MainPage(),
    MainPage(),
    MainPage(),
  ];

... 


bottomNavigationBar: Theme(
            data: Theme.of(context).copyWith(
              canvasColor: Color(0xFF3B3D58),
              primaryColor: Colors.white,
              textTheme: Theme.of(context).textTheme.copyWith(
                caption: TextStyle(color: Colors.grey)
              )
            ),
            child: BottomNavigationBar(
              type: BottomNavigationBarType.fixed,
              currentIndex: _page,
              items: [
                BottomNavigationBarItem(
                  icon: Icon(Icons.home),
                  title: Text('Home'),
                ),
                BottomNavigationBarItem(
                  icon: Icon(Icons.add_circle),
                  title: Text('Add Place'),
                ),
                BottomNavigationBarItem(
                  icon: Icon(Icons.near_me),
                  title: Text('Near Me'),
                ),
                BottomNavigationBarItem(
                  icon: Icon(Icons.favorite),
                  title: Text('Favorite'),
                ),
                BottomNavigationBarItem(
                  icon: Icon(Icons.more_horiz),
                  title: Text('More'),
                ),
              ],
              onTap: onTabTapped,
            ),
          ),
```

And this is my result of my code:
[Here](https://i.stack.imgur.com/SaUz1.png)

**Comment by bhargav6744** [comment]
` BottomNavigationBarItem(
              backgroundColor: Colors.red,)`

I tried this but even it isn't working

**Comment by rrifafauzikomara** [comment]
@bhargav6744 Yes, I already try it but it isn't working

**Comment by tony123S** [comment]
doesn't work for me too

**Comment by Surjitpal** [comment]
Hi team use below property with for background color

backgroundColor: Colors.green,
type: BottomNavigationBarType.fixed,

**Comment by sabsab-yeng** [comment]
thanks is working for me

**Comment by sabsab-yeng** [comment]
bottomNavigationBar: Theme(
        data: Theme.of(context).copyWith(
          canvasColor: appbarColor,
          primaryColor: appColor,
          textTheme: Theme.of(context)
              .textTheme
              .copyWith(caption: TextStyle(color: Colors.black)),
        ),
        child: BottomNavigationBar(
          currentIndex: _curentIndex,
          onTap: onTapp,
          items: [
            BottomNavigationBarItem(
              icon: Icon(
                Icons.home,
              ),
              title: Text("Home"),
            ),
            BottomNavigationBarItem(
              icon: Icon(
                Icons.history,
              ),
              title: Text("History"),
            ),
            BottomNavigationBarItem(
              icon: Icon(
                Icons.flight,
              ),
              title: Text("Cows"),
            ),
          ],
        ),
      ),

**Comment by PhamTruongHung** [comment]
Thank you, that's worked!

**Comment by DanielJaramillo94** [comment]
For now, i just replaced the actual implementation for the one gived on pull request https://github.com/flutter/flutter/pull/53261 :/

For example, this is the implementation for Material package
https://raw.githubusercontent.com/digitalkaoz/flutter/fe7bb851d5de69db97dcdc297cc779c771804a57/packages/flutter/lib/src/material/bottom_navigation_bar.dart

**Comment by rohit-karthik** [comment]
I believe you just use 
`selectedItemColor` and
`unselectedItemColor` in the `BottomNavigationBar()` widget contents. These will allow you to change the color for each Bottom Navigation Bar item. Example

`bottomNavigationBar: BottomNavigationBar(
            backgroundColor: Colors.blueAccent,
            selectedItemColor: Colors.white,
            unselectedItemColor: Colors.black,
            items: [
              BottomNavigationBarItem(
                icon: Icon(
                  Icons.category,
                ),
                label: "Test",
              ),
            ],
          ),`

**Comment by AyushBherwani1998** [comment]
I would like to work on the issue, we also have a requirement to show gradient when `BottomNavigationBarItem` is selected. Is it something we are also looking to added? 
cc:\\ @shihaohong @HansMuller

**Comment by shihaohong** [comment]
@AyushBherwani1998 By gradient, did you mean to have the color value lerp between selected/unselected states?

**Comment by AyushBherwani1998** [comment]
No, actually instead of having a solid background, we have a requirement for the gradient.  For reference, I'm attaching a image below.
![Capture](https://user-images.githubusercontent.com/34301187/100883465-6067e780-34d6-11eb-960e-62bf206ff6e8.PNG)

**Comment by AyushBherwani1998** [comment]
Hey, @shihaohong I did check out the implementation, my first thoughts are to change the `backgroundColor` property from `Color` to `MaterialStateProperty<Color>`. This way we will be handling the `MaterialState.selected` and default color for the unselected state.

We can add a new property of `Color? background` to class [`_BottomNavigatonTile`](https://github.com/flutter/flutter/blob/78910062997c3a836feee883712c241a5fd22983/packages/flutter/lib/src/material/bottom_navigation_bar.dart#L337) and while adding the `_BottomNavigationTile` we can pass background color after resolving states. I did some quick changes, and it works fine.

<img src="https://user-images.githubusercontent.com/34301187/102006875-58b40880-3d4a-11eb-8af2-fd082ccbd629.png" width=200>

Since, I'm adding the property, should we also deprecated `selectedItemColor` and `unselectedItemColor` and add `MaterialStateProperty<Color> itemColor`. The same suggestion goes for other similar properties too.

**Comment by AyushBherwani1998** [comment]
@shihaohong just a reminder, this will be breaking change if we just change the `background` from `Color` to `MaterialStateProperty<Color>`. Can we introduce a new property and deprecate the `backgroundColor`?

**Comment by shihaohong** [comment]
@AyushBherwani1998 I don't think it has to be a breaking change (or maybe not yet) -- We can have those properties coexist, and if the passed in color is a `MaterialStateProperty`, to simply ignore `selectedItemColor` and `unselectedItemColor`. We would document this behavior as well. Would it be possible to do this with `backgroundColor` and not create a new property?

This was what I was hoping to do with ToggleButtons if I get back to it: https://github.com/flutter/flutter/issues/36861.

**Comment by adityaguptagkp30** [comment]
> I want every menu that I click will change color for each menu that is clicked, not for the entire color background of the nav bar.
> 
> Like the example below:
> [Here](https://i.stack.imgur.com/dXGaz.png)
> 
> When a user clicks on the "Near Me" menu then the box on the menu will change color. But for the background/color box on the other menu, it is still black. Similarly, clicking on the other menus will change color to red and will not change the color of the menu that is not clicked (still black)
> 
> This my Code:
> 
> ```
> int _page = 0;
> final List<Widget> _children = [
>     MainPage(),
>     MainPage(),
>     MainPage(),
>     MainPage(),
>     MainPage(),
>   ];
> 
> ... 
> 
> 
> bottomNavigationBar: Theme(
>             data: Theme.of(context).copyWith(
>               canvasColor: Color(0xFF3B3D58),
>               primaryColor: Colors.white,
>               textTheme: Theme.of(context).textTheme.copyWith(
>                 caption: TextStyle(color: Colors.grey)
>               )
>             ),
>             child: BottomNavigationBar(
>               type: BottomNavigationBarType.fixed,
>               currentIndex: _page,
>               items: [
>                 BottomNavigationBarItem(
>                   icon: Icon(Icons.home),
>                   title: Text('Home'),
>                 ),
>                 BottomNavigationBarItem(
>                   icon: Icon(Icons.add_circle),
>                   title: Text('Add Place'),
>                 ),
>                 BottomNavigationBarItem(
>                   icon: Icon(Icons.near_me),
>                   title: Text('Near Me'),
>                 ),
>                 BottomNavigationBarItem(
>                   icon: Icon(Icons.favorite),
>                   title: Text('Favorite'),
>                 ),
>                 BottomNavigationBarItem(
>                   icon: Icon(Icons.more_horiz),
>                   title: Text('More'),
>                 ),
>               ],
>               onTap: onTabTapped,
>             ),
>           ),
> ```
> 
> And this is my result of my code:
> [Here](https://i.stack.imgur.com/SaUz1.png)

where is your ontapped function

**Comment by vytautas-pranskunas-** [comment]
Any updates on this? Cannot believe it is pending for so long... :(

**Comment by EbramTawfik** [comment]
I found this workaround:

https://dartpad.dev/?id=60405c72383f2114decc8486b427e03e&null_safety=true

**Comment by adityaguptagkp30** [comment]
> Any updates on this? Cannot believe it is pending for so long... :(

Yes it's done thanks for replying

**Comment by EbramTawfik** [comment]
@adityaguptagkp30 

> > Any updates on this? Cannot believe it is pending for so long... :(
> 
> Yes it's done thanks for replying

It's done but not released yet?

**Comment by Jedibooney** [comment]
Has there been any permanent, released fix for this?

**Comment by trapeye** [comment]
> Has there been any permanent, released fix for this?

you know how to fix it?

**Comment by EArminjon** [comment]
When BottomNavigationBar have `BottomNavigationBarType.shifting` the background color use if the BottomNavigationBarItem's background color. When BottomNavigationBar have  `BottomNavigationBarType.shifting` the BottomNavigationBar's background color is used.

**Comment by vietstone-ng** [comment]
> I found this workaround: Try here: https://dartpad.dev/?id=60405c72383f2114decc8486b427e03e&null_safety=true
> 
> code here:
> 
> ```
> import 'package:flutter/material.dart';
> 
> 
> void main() => runApp(const MyApp());
> 
> /// This is the main application widget.
> class MyApp extends StatelessWidget {
>   const MyApp({Key? key}) : super(key: key);
> 
>   static const String _title = 'Flutter Code Sample';
> 
>   @override
>   Widget build(BuildContext context) {
>     return const MaterialApp(
>       title: _title,
>       home: MyStatefulWidget(),
>     );
>   }
> }
> 
> /// This is the stateful widget that the main application instantiates.
> class MyStatefulWidget extends StatefulWidget {
>   const MyStatefulWidget({Key? key}) : super(key: key);
> 
>   @override
>   State<MyStatefulWidget> createState() => _MyStatefulWidgetState();
> }
> 
> /// This is the private State class that goes with MyStatefulWidget.
> class _MyStatefulWidgetState extends State<MyStatefulWidget> {
>   int _selectedIndex = 0;
>   static const TextStyle optionStyle =
>       TextStyle(fontSize: 30, fontWeight: FontWeight.bold);
>   static const List<Widget> _widgetOptions = <Widget>[
>     Text(
>       'Index 0: Home',
>       style: optionStyle,
>     ),
>     Text(
>       'Index 1: Business',
>       style: optionStyle,
>     ),
>     Text(
>       'Index 2: School',
>       style: optionStyle,
>     ),
>   ];
> 
>   void _onItemTapped(int index) {
>     setState(() {
>       _selectedIndex = index;
>     });
>   }
> 
>   @override
>   Widget build(BuildContext context) {
>     return Scaffold(
>       appBar: AppBar(
>         title: const Text('BottomNavigationBar Sample'),
>       ),
>       body: Center(
>         child: _widgetOptions.elementAt(_selectedIndex),
>       ),
>       bottomNavigationBar: BottomNavigationBar(
>         items: <BottomNavigationBarItem>[
>           _bottomNavigationBarItem(Icons.home, 'Home'),
>           _bottomNavigationBarItem(Icons.business, 'Home'),
>           _bottomNavigationBarItem(Icons.school, 'Home'),
>         ],
>         currentIndex: _selectedIndex,
>         selectedItemColor: Colors.amber[800],
>         elevation: 0,
>         onTap: _onItemTapped,
>       ),
>     );
>   }
> 
>   BottomNavigationBarItem _bottomNavigationBarItem(
>       IconData icon, String label) {
>     return BottomNavigationBarItem(
>       activeIcon: _navItemIcon(icon, label, Colors.red, Colors.white),
>       icon: _navItemIcon(icon, label, Color(0xff3b3c58), Colors.grey),
>       title: Padding(padding: EdgeInsets.all(0)),
>     );
>   }
> 
>   Row _navItemIcon(IconData icon, String label, Color? backgrondColor,
>       Color? foregroundColor) {
>     return Row(
>       children: [
>         Expanded(
>           child: Container(
>             color: backgrondColor,
>             child: Padding(
>               padding: const EdgeInsets.all(12.0),
>               child: Column(
>                 children: [
>                   Icon(
>                     icon,
>                     color: foregroundColor,
>                   ),
>                   Text(
>                     label,
>                     style: TextStyle(color: foregroundColor),
>                   )
>                 ],
>               ),
>             ),
>           ),
>         ),
>       ],
>     );
>   }
> }
> ```

Seems promising. But not work for now, as `title` widget was removed, and `label` string was added, and required :(
